src/test/java/software/amazon/encryption/s3/S3EncryptionClientReEncryptInstructionFileTest.java

```diff
@@ -0,0 +1,1270 @@
+
+    JsonNode clientInstructionFileNode = parser.parse(clientInstructionFileContent);
+    String clientIv = clientInstructionFileNode.asObject().get("x-amz-iv").asString();
+    String clientEncryptedDataKeyAlgorithm = clientInstructionFileNode.asObject().get("x-amz-wrap-alg").asString();
+    String clientEncryptedDataKey = clientInstructionFileNode.asObject().get("x-amz-key-v2").asString();
+    JsonNode clientMatDescNode = parser.parse(clientInstructionFileNode.asObject().get("x-amz-matdesc").asString());
+
+    assertEquals("yes", clientMatDescNode.asObject().get("isOwner").asString());
+    assertEquals("admin", clientMatDescNode.asObject().get("access-level").asString());
+
+    ResponseBytes<GetObjectResponse> thirdPartyInstFile = wrappedClient.getObjectAsBytes(builder -> builder
+      .bucket(BUCKET)
+      .key(objectKey + ".third-party-access-instruction-file")
+      .build());
+
+    String thirdPartyInstructionFileContent = thirdPartyInstFile.asUtf8String();
+    JsonNode thirdPartyInstructionFileNode = parser.parse(thirdPartyInstructionFileContent);
+    String thirdPartyIv = thirdPartyInstructionFileNode.asObject().get("x-amz-iv").asString();
+    String thirdPartyEncryptedDataKeyAlgorithm = thirdPartyInstructionFileNode.asObject().get("x-amz-wrap-alg").asString();
+    String thirdPartyEncryptedDataKey = thirdPartyInstructionFileNode.asObject().get("x-amz-key-v2").asString();
+    JsonNode thirdPartyMatDescNode = parser.parse(thirdPartyInstructionFileNode.asObject().get("x-amz-matdesc").asString());
+    assertEquals("no", thirdPartyMatDescNode.asObject().get("isOwner").asString());
+    assertEquals("user", thirdPartyMatDescNode.asObject().get("access-level").asString());
+
+    assertEquals(clientIv, thirdPartyIv);
+    assertEquals(clientEncryptedDataKeyAlgorithm, thirdPartyEncryptedDataKeyAlgorithm);
+    assertNotEquals(clientEncryptedDataKey, thirdPartyEncryptedDataKey);
+
+    try {
+      ResponseBytes<GetObjectResponse> thirdPartyDecryptObject = thirdPartyClient.getObjectAsBytes(builder -> builder
+        .bucket(BUCKET)
+        .key(objectKey)
+        .build());
+      throw new RuntimeException("Expected exception");
+    } catch (S3EncryptionClientException e) {
+      assertTrue(e.getMessage().contains("Unable to RSA-OAEP-SHA1 unwrap"));
+    }
+
+    ResponseBytes<GetObjectResponse> thirdPartyDecryptedObject = thirdPartyClient.getObjectAsBytes(builder -> builder
+      .bucket(BUCKET)
+      .key(objectKey)
+      .overrideConfiguration(withCustomInstructionFileSuffix(".third-party-access-instruction-file"))
+      .build());
+    assertEquals(input, thirdPartyDecryptedObject.asUtf8String());
+
+    ResponseBytes<GetObjectResponse> clientDecryptedObject = client.getObjectAsBytes(builder -> builder
+      .bucket(BUCKET)
+      .key(objectKey)
+      .build());
+    assertEquals(input, clientDecryptedObject.asUtf8String());
+
+    assertEquals(BUCKET, reEncryptInstructionFileResponse.Bucket());
```

Comment:
probably move these to after the `reEncryptInstructionFile` call so it's more readable.